### PR TITLE
chore: filter scan job details to avoid api-server resource overflow

### PR DESCRIPTION
**Pull request by erikgb**
I suggest creating scan jobs that don't use `minSeverity` and `ignoreUnfixed` as input to the scan job, but instead apply these filters to the results of the scan job. This should improve the handling of "overflow" better, and avoid the ping-pong between controllers.

This is the first step, improving the handling of api-server resource overflow. After this PR, we just filter the vulnerability report by severity until it fits into the api-server. This simplifies the machinery and allows us to take back the currently disabled vulnerability overflow e2e-test.

A future improvement could be to avoid creating multiple scan jobs for the same image.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -222,8 +222,7 @@ $(GCI):
 ##@ End-to-end (e2e) Testing
 
 e2e-test: ## Run e2e tests
-	chainsaw test --test-dir test/e2e/scenario --config test/e2e-config/.chainsaw.yaml
-	chainsaw test --test-dir test/e2e/workload-scan --config test/e2e-config/.chainsaw.yaml
+	chainsaw test --test-dir test/e2e/scenario --test-dir test/e2e/workload-scan --config test/e2e-config/.chainsaw.yaml
```
Comment (tenstad) [inline Makefile:225]: Could do the chainsaw stuff in a separate PR

Comment (erikgb) [reply Makefile:225]: It is quite related, so I think it's fine to have it in the same PR.

**Comment by tenstad** [comment]
Tempting to also do the `string` -> `SeverityName` in a separate refactor PR. WDYT?

**Comment by erikgb** [comment]
> Tempting to also do the `string` -> `SeverityName` in a separate refactor PR. WDYT?

Then we would have to change things that are going to be deleted. We could avoid changing it in this PR, tho. And maybe do it later on.

File: internal/controller/stas/scan_job_controller.go [hunk 2]
```diff
@@ -389,9 +380,24 @@ func (r *ScanJobReconciler) getScanJobLogs(ctx context.Context, job *batchv1.Job
 	return r.GetLogs(ctx, client.ObjectKeyFromObject(&jobPod), trivy.ScanJobContainerName)
 }
 
+func filterVulnerabilities(orig []stasv1alpha1.Vulnerability, minSeverity stasv1alpha1.Severity) (filtered []stasv1alpha1.Vulnerability, err error) {
+	for _, v := range orig {
+		severity, err := stasv1alpha1.NewSeverity(v.Severity)
+		if err != nil {
+			return filtered, err
```
Comment (tenstad) [inline internal/controller/stas/scan_job_controller.go:387]:
```suggestion
			return nil, err
```

File: internal/controller/stas/scan_job_controller.go [hunk 3]
```diff
@@ -389,9 +380,24 @@ func (r *ScanJobReconciler) getScanJobLogs(ctx context.Context, job *batchv1.Job
 	return r.GetLogs(ctx, client.ObjectKeyFromObject(&jobPod), trivy.ScanJobContainerName)
 }
 
+func filterVulnerabilities(orig []stasv1alpha1.Vulnerability, minSeverity stasv1alpha1.Severity) (filtered []stasv1alpha1.Vulnerability, err error) {
```
Comment (tenstad) [inline internal/controller/stas/scan_job_controller.go:383]: Not that fan of named return params. Are we using it elsewhere, or introducing it now?

File: internal/controller/stas/scan_job_controller.go [hunk 4]
```diff
@@ -202,28 +201,20 @@ func (r *ScanJobReconciler) reconcileCompleteJob(ctx context.Context, job *batch
...
-	err = r.Status().Patch(ctx, cis, client.MergeFrom(cleanCis))
-	if err != nil && isResourceTooLargeError(err) {
-		cis = cleanCis.DeepCopy()
-
-		condition := metav1.Condition{
-			Type:    string(kstatus.ConditionStalled),
-			Status:  metav1.ConditionTrue,
-			Reason:  stasv1alpha1.ReasonVulnerabilityOverflow,
-			Message: fmt.Sprintf("vulnerability report is too large to fit in API: %s", err),
+	for {
```
Comment (tenstad) [inline internal/controller/stas/scan_job_controller.go:204]: Exit condition is quite hidden without it
```suggestion
	for minSeverity < stasv1alpha1.MaxSeverity {
```

Comment (tenstad) [reply internal/controller/stas/scan_job_controller.go:204]: or 
```suggestion
	for filterSeverity := minSeverity ; i < stasv1alpha1.MaxSeverity;  filterSeverity++ {
```

File: internal/controller/stas/scan_job_controller.go (hunk 3 above)
Comment (erikgb) [reply internal/controller/stas/scan_job_controller.go:383]: Agreed, I copied it from an example I found. I'll fix it.

File: internal/controller/stas/scan_job_controller.go (hunk 4 above)
Comment (erikgb) [reply internal/controller/stas/scan_job_controller.go]: Fixed

File: internal/controller/stas/scan_job_controller.go (hunk 3 above)
Comment (erikgb) [reply internal/controller/stas/scan_job_controller.go]: Fixed

File: internal/controller/stas/scan_job_controller.go [hunk 5]
```diff
@@ -202,24 +201,16 @@ func (r *ScanJobReconciler) reconcileCompleteJob(ctx context.Context, job *batch
...
-	if err != nil && isResourceTooLargeError(err) {
-		cis = cleanCis.DeepCopy()
-
-		condition := metav1.Condition{
-			Type:    string(kstatus.ConditionStalled),
-			Status:  metav1.ConditionTrue,
-			Reason:  stasv1alpha1.ReasonVulnerabilityOverflow,
-			Message: fmt.Sprintf("vulnerability report is too large to fit in API: %s", err),
+	// Repeat until resource fits in api-server by increasing minimum severity on failure.
+	for severity := minSeverity; severity < stasv1alpha1.MaxSeverity; severity++ {
```
Comment (tenstad) [inline internal/controller/stas/scan_job_controller.go:205]: ?
```suggestion
	for severity := minSeverity; severity <= stasv1alpha1.MaxSeverity; severity++ {
```

Comment (erikgb) [reply internal/controller/stas/scan_job_controller.go]: Thanks! Good catch!

**Review by tenstad — APPROVED** [review]
